examples/ex31p_proposed.cpp

```diff
@@ -0,0 +1,530 @@
+
+//                       MFEM Example 31 - Parallel Version
+//
+// Compile with: make ex31p
+//
+// Sample runs:  mpirun -np 4 ex31p -m ../data/hexagon.mesh -o 2
+//               mpirun -np 4 ex31p -m ../data/star.mesh
+//               mpirun -np 4 ex31p -m ../data/square-disc.mesh -o 2 -n 4 -rs 1
```

Comment:
Suspicious results when running `mpirun -np 4 ex31p_proposed -m ../data/square-disc.mesh -o 2 -n 4 -rs 1`
<img width="802" alt="Screen Shot 2021-05-06 at 12 01 10 PM" src="https://user-images.githubusercontent.com/11756601/117351580-c0d2ca80-ae62-11eb-8b57-582873a217be.png">

Reply:
Is the genus of the domain being properly handled here? The cohomology is non-trivial here....

Reply:
> Not that different results are produced from `mpirun -np 1 ex31p_proposed -m ../data/square-disc.mesh -o 2 -n 4 -rs 1`

This also could be related to https://github.com/mfem/mfem/issues/2226. Thanks @psocratis 
<img width="799" alt="Screen Shot 2021-05-06 at 12 56 56 PM" src="https://user-images.githubusercontent.com/11756601/117358052-8c630c80-ae6a-11eb-87d5-c5cc5330e8b4.png">

Reply:
![GLVis_s01](https://user-images.githubusercontent.com/6109571/117361494-f4b3ed00-ae6e-11eb-9ae5-8557fd3afe75.png)
This result looks much better on the NURBs mesh. In any case this is certainly a good question. HypreAME should be using the discrete gradient operator to weed out gradients of scalar fields. This should leave the rotational fields we're interested in and a single irrotational field due to the hole in the topology. This irrotational field doesn't wrap around the hole as I would expect but I cannot say if it's correct or not.

As for the other three panels; they look ugly. I was considering rescaling the panels so that the XY panel and the Z panel use the same color bar. I think this would help to alleviate some of the confusion caused by this set of images.

Do you think this rescaling is a good idea in general? Should we avoid this topology in particular? Should we switch to the NURBs mesh (I think we should)?

Reply:
It's really interesting that the results are so much poorer with the high order mesh than with the NURBS mesh. I would suggest that the default behaviour be not to plot the zero eigenmodes. That would bypass the need for any rescaling.

Reply:
It may be good to keep this there, alongside the NURBS mesh example. There is some implicit information in there. If someone runs both, they may decide right away that it's worth using a NURBS description of their domain geometry.

Reply:
> ![GLVis_s01](https://user-images.githubusercontent.com/6109571/117361494-f4b3ed00-ae6e-11eb-9ae5-8557fd3afe75.png)
> This result looks much better on the NURBs mesh. In any case this is certainly a good question. HypreAME should be using the discrete gradient operator to weed out gradients of scalar fields. This should leave the rotational fields we're interested in and a single irrotational field due to the hole in the topology. This irrotational field doesn't wrap around the hole as I would expect but I cannot say if it's correct or not.
> 
> As for the other three panels; they look ugly. I was considering rescaling the panels so that the XY panel and the Z panel use the same color bar. I think this would help to alleviate some of the confusion caused by this set of images.
> 
> Do you think this rescaling is a good idea in general? Should we avoid this topology in particular? Should we switch to the NURBs mesh (I think we should)?

@mlstowell Can you share the command you used to produce this figure? I am not seeing such good results with `../data/square-disc-nurbs.mesh` anymore :/

Reply:
For example, `❯ mpirun -np 4 ex31p_proposed -m ../data/square-disc-nurbs.mesh -o 2 -n 4 -rs 1` produces
<img width="244" alt="Screen Shot 2021-05-18 at 10 53 20 AM" src="https://user-images.githubusercontent.com/11756601/118700208-4477aa00-b7c7-11eb-9421-6698a4fc1a65.png">

Reply:
I don't recall exactly but `./ex31p_proposed -m ../data/square-disc-nurbs.mesh -rs 4 -o 2` looks quite similar. This one's not bad either `./ex31p_proposed -m ../data/square-disc-nurbs.mesh -rs 3 -o 3`.

Reply:
OK. Perhaps the default should be `./ex31p_proposed -m ../data/square-disc-nurbs.mesh -rs 3 -o 3`
`mpirun -np 4 ex31p_proposed -m ../data/square-disc-nurbs.mesh` doesn't look very nice